cryptonight/src/hash_v2.rs

```diff
@@ -0,0 +1,445 @@
+    // see: https://github.com/monero-project/monero/blob/v0.18.3.3/src/crypto/variant2_int_sqrt.h#L65-L152
+    let sqrt_div2 = sqrt_result >> 1;
+    let lsb = sqrt_result & 1;
+    let r2 = sqrt_div2
+        .wrapping_mul(sqrt_div2 + lsb)
+        .wrapping_add(sqrt_result << 32);
+
+    if r2.wrapping_add(lsb) > sqrt_input {
+        sqrt_result = sqrt_result.wrapping_sub(1);
+    }
+    if r2.wrapping_add(1 << 32) < sqrt_input.wrapping_sub(sqrt_div2) {
+        // Not sure that this is possible. I tried writing a test program
+        // to search subsets of u64 for a value that can trigger this
+        // branch, but couldn't find anything. The Go implementation came
+        // to the same conclusion:
+        // https://github.com/Equim-chan/cryptonight/blob/v0.3.0/arith_ref.go#L39-L45
+        sqrt_result = sqrt_result.wrapping_add(1);
+    }
+
+    sqrt_result
+}
+
+// Original C code:
+// https://github.com/monero-project/monero/blob/v0.18.3.4/src/crypto/slow-hash.c#L277-L283
+pub(crate) fn variant2_integer_math(
+    c2: &mut [u8; 8],
+    c1: &[u8; AES_BLOCK_SIZE],
+    division_result: &mut u64,
+    sqrt_result: &mut u64,
+    variant: Variant,
+) {
+    const U32_MASK: u64 = u32::MAX as u64;
+
+    if variant == Variant::V2 {
+        let tmpx = *division_result ^ (*sqrt_result << 32);
+        *c2 = (u64::from_le_bytes(*c2) ^ tmpx).to_le_bytes();
+
+        let c1_64 = block_to_u64le(c1);
+        let mut divisor = c1_64[0];
+        let dividend = c1_64[1];
+
+        divisor = ((divisor + ((*sqrt_result << 1) & U32_MASK)) | 0x80000001) & U32_MASK;
+        *division_result =
+            ((dividend / divisor) & U32_MASK).wrapping_add((dividend % divisor) << 32);
+
+        let sqrt_input = c1_64[0].wrapping_add(*division_result);
+        *sqrt_result = variant2_integer_math_sqrt(sqrt_input);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use digest::Digest;
+    use groestl::Groestl256;
+
+    use super::*;
+    use crate::util::{hex_to_array, subarray_mut};
+
+    #[test]
+    fn test_variant2_integer_math() {
+        let test = |c2_hex: &str,
+                    c1_hex: &str,
+                    division_result: u64,
+                    sqrt_result: u64,
+                    c2_hex_end: &str,
+                    division_result_end: u64,
+                    sqrt_result_end: u64| {
+            let mut c2: [u8; 16] = hex_to_array(c2_hex);
+            let c1: [u8; 16] = hex_to_array(c1_hex);
+            let mut division_result = division_result;
+            let mut sqrt_result = sqrt_result;
+
+            variant2_integer_math(
+                subarray_mut(&mut c2, 0),
+                &c1,
+                &mut division_result,
+                &mut sqrt_result,
+                Variant::V2,
+            );
+
+            assert_eq!(hex::encode(c2), c2_hex_end);
+            assert_eq!(division_result, division_result_end);
+            assert_eq!(sqrt_result, sqrt_result_end);
+        };
+        test(
```

Comment:
Non-capturing closure can just be a `fn`.
```suggestion
        fn test(
            c2_hex: &str,
            c1_hex: &str,
            division_result: u64,
            sqrt_result: u64,
            c2_hex_end: &str,
            division_result_end: u64,
            sqrt_result_end: u64,
        ) {
            let mut c2: [u8; 16] = hex_to_array(c2_hex);
            let c1: [u8; 16] = hex_to_array(c1_hex);
            let mut division_result = division_result;
            let mut sqrt_result = sqrt_result;

            variant2_integer_math(
                subarray_mut(&mut c2, 0),
                &c1,
                &mut division_result,
                &mut sqrt_result,
                Variant::V2,
            );

            assert_eq!(hex::encode(c2), c2_hex_end);
            assert_eq!(division_result, division_result_end);
            assert_eq!(sqrt_result, sqrt_result_end);
        }

        test(
```